### PR TITLE
feat(db): Refactor storage libraries

### PR DESCRIPTION
**Pull request by rakita**
related to this: https://github.com/paradigmxyz/reth/issues/333

The only difference is that provider is made standalone as in `storage/reth-provider`

### CODE REVIEW

**Comment by gakonst** [comment]
Some import errors around the macros i suppose

File: crates/storage/db/src/tables/codecs/postcard.rs [hunk 1]
```diff
@@ -36,5 +39,5 @@ macro_rules! impl_postcard {
 
 type VecU8 = Vec<u8>;
 
-#[cfg(feature = "bench-postcard")]
-impl_postcard!(VecU8, Receipt, H256, U256, H160, u8, u16, u64, Header, Account, Log, TxType);
+//#[cfg(feature = "bench-postcard")]
```
Comment (rakita) [inline crates/storage/db/src/tables/codecs/postcard.rs:42]: @joshieDo I removed postcard as it clashes with impl of Decode/Encode from table.

File: crates/storage/db/src/tables/codecs/mod.rs [hunk 2]
```diff
@@ -4,6 +4,6 @@ mod compact;
 pub use compact::CompactU256;
 
 pub mod fuzz;
+
 mod postcard;
-#[cfg(not(feature = "bench-postcard"))]
 mod scale;
```
Comment (rakita) [inline crates/storage/db/src/tables/codecs/mod.rs:9]: And removed this `cfg` as scale implements some of our `Compress/Decompress` traits

File: crates/storage/db/src/tables/codecs/postcard.rs (hunk 1 above)
Comment (gakonst) [reply crates/storage/db/src/tables/codecs/postcard.rs:42]: Wdym it crashes?

**Review by gakonst — APPROVED** [review]

File: crates/storage/db/src/tables/codecs/postcard.rs (hunk 1 above)
Comment (rakita) [reply crates/storage/db/src/tables/codecs/postcard.rs:42]: `Decode/Encode` are implemented in two places. 
One in tables:
https://github.com/paradigmxyz/reth/blob/e014eaad339e4b01f6c682614a15c0c0651cc218/crates/storage/db/src/tables/models/mod.rs#L19-L46
And second one is for postcard:
https://github.com/paradigmxyz/reth/blob/e014eaad339e4b01f6c682614a15c0c0651cc218/crates/storage/db/src/tables/codecs/postcard.rs#L20-L43

When I thnk about it now, postcard probably needs to implement `Compress/Decompress` and not Encode/Decode.

Comment (gakonst) [reply crates/storage/db/src/tables/codecs/postcard.rs:42]: Interesting. Honestly having both of these traits is kinda confusing. @joshieDo ideas on how we can improve this?

Comment (joshieDo) [reply crates/storage/db/src/tables/codecs/postcard.rs:42]: > When I thnk about it now, postcard probably needs to implement Compress/Decompress and not Encode/Decode.

Hm, yeah when we made the change to `Compress/Decompress`, we didn't change these.

**Pull request merged**